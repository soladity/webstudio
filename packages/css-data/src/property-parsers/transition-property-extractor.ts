import type {
  KeywordValue,
  TupleValue,
  UnitValue,
} from "@webstudio-is/css-engine";

export const isTimingFunction = (timing: string) => {
  const regex =
    /^(ease(-in-out|-in|-out)?|linear|cubic-bezier\((-?\d+(\.\d+)?, ?){3}-?\d+(\.\d+)?\)|steps\(\d+(-?(start|end))?\))$/gm;
  return regex.test(timing);
};

export const extractTransitionProperties = (
  transition: TupleValue
): {
  property: KeywordValue | null;
  timing: KeywordValue | null;
  duration: UnitValue | null;
  delay: UnitValue | null;
} => {
  let property: KeywordValue | null = null;
  let timing: KeywordValue | null = null;
  let duration: UnitValue | null = null;
  let delay: UnitValue | null = null;

  const keywordValues: KeywordValue[] = [];
  const unitValues: UnitValue[] = [];

  for (const property of transition.value) {
    if (property.type === "keyword") {
      keywordValues.push(property);
    }

    if (property.type === "unit") {
      unitValues.push(property);
    }
  }

  if (keywordValues.length && isTimingFunction(keywordValues[0].value)) {
    timing = keywordValues[0];
  } else {
    property = keywordValues[0];
    timing = keywordValues[1] ?? null;
  }

  duration = unitValues[0] ?? null;
  delay = unitValues[1] ?? null;

  return {
    property,
    duration,
    timing,
    delay,
  };
};
