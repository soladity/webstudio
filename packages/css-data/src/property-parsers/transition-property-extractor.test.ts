import { describe, expect, test } from "@jest/globals";
import { extractTransitionProperties } from "./transition-property-extractor";

describe("extractTransitionProperties", () => {
  test("parses and extracts values from opacity 200ms ease 50ms", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "unit",
          value: 200,
          unit: "ms",
        },
        {
          type: "keyword",
          value: "ease",
        },
        {
          type: "unit",
          value: 50,
          unit: "ms",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: { type: "unit", value: 200, unit: "ms" },
      timing: { type: "keyword", value: "ease" },
      delay: { type: "unit", value: 50, unit: "ms" },
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": {
          "type": "unit",
          "value": 50,
          "unit": "ms"
        }
      }
      `);
  });

  test("parses and extracts values from opacity 200ms ease", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "unit",
          value: 200,
          unit: "ms",
        },
        {
          type: "keyword",
          value: "ease",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: { type: "unit", value: 200, unit: "ms" },
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
      `);
  });

  test("parses and extracts values from opacity 200ms 10ms", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "unit",
          value: 200,
          unit: "ms",
        },
        {
          type: "unit",
          value: 10,
          unit: "ms",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: { type: "unit", value: 200, unit: "ms" },
      timing: null,
      delay: { type: "unit", value: 10, unit: "ms" },
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": null,
        "delay": {
          "type": "unit",
          "value": 10,
          "unit": "ms"
        }
      }
    `);
  });

  test("parses and extracts values from opacity 200ms", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "unit",
          value: 200,
          unit: "ms",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: { type: "unit", value: 200, unit: "ms" },
      timing: null,
      delay: null,
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": null,
        "delay": null
      }
      `);
  });

  test("parses and extracts values from ease 10ms", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "ease",
        },
        {
          type: "unit",
          value: 10,
          unit: "ms",
        },
      ],
    });
    expect(result).toEqual({
      property: null,
      duration: { type: "unit", value: 10, unit: "ms" },
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });
  });

  test("parses and extracts values from opacity ease", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "keyword",
          value: "ease",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: null,
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
    `);
  });

  test("parses and extracts values from opacity ease 50ms", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "opacity",
        },
        {
          type: "keyword",
          value: "ease",
        },
        {
          type: "unit",
          value: 50,
          unit: "ms",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "opacity" },
      duration: { type: "unit", value: 50, unit: "ms" },
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 50,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
    `);
  });

  test("parses and extracts values from transform 0.6s cubic-bezier(0.36, 0, 0.66, -0.56)", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "transform",
        },
        {
          type: "unit",
          value: 0.6,
          unit: "s",
        },
        {
          type: "keyword",
          value: "cubic-bezier(0.36, 0, 0.66, -0.56)",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "transform" },
      duration: { type: "unit", value: 0.6, unit: "s" },
      timing: { type: "keyword", value: "cubic-bezier(0.36, 0, 0.66, -0.56)" },
      delay: null,
    });
    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "transform"
        },
        "duration": {
          "type": "unit",
          "value": 0.6,
          "unit": "s"
        },
        "timing": {
          "type": "keyword",
          "value": "cubic-bezier(0.36, 0, 0.66, -0.56)"
        },
        "delay": null
      }
    `);
  });

  test("parses and extracts values from custom-property ease", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "custom-property",
        },
        {
          type: "keyword",
          value: "ease",
        },
      ],
    });

    expect(result).toEqual({
      property: { type: "keyword", value: "custom-property" },
      duration: null,
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });

    expect(result).toMatchSnapshot(`
      {
        "property": {
          "type": "keyword",
          "value": "custom-property"
        },
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
    `);
  });

  test("parses and extracts values from custom-property ease", () => {
    const result = extractTransitionProperties({
      type: "tuple",
      value: [
        {
          type: "keyword",
          value: "ease",
        },
      ],
    });

    expect(result).toEqual({
      property: null,
      duration: null,
      timing: { type: "keyword", value: "ease" },
      delay: null,
    });

    expect(result).toMatchSnapshot(`
      {
        "property": null,
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
    `);
  });
});
