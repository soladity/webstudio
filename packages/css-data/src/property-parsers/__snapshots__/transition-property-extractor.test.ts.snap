// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTransitionProperties parses and extracts values from custom-property ease: 
      {
        "property": {
          "type": "keyword",
          "value": "custom-property"
        },
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
     1`] = `
{
  "delay": null,
  "duration": null,
  "property": {
    "type": "keyword",
    "value": "custom-property",
  },
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from custom-property ease: 
      {
        "property": null,
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
     1`] = `
{
  "delay": null,
  "duration": null,
  "property": null,
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity 200ms 10ms: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": null,
        "delay": {
          "type": "unit",
          "value": 10,
          "unit": "ms"
        }
      }
     1`] = `
{
  "delay": {
    "type": "unit",
    "unit": "ms",
    "value": 10,
  },
  "duration": {
    "type": "unit",
    "unit": "ms",
    "value": 200,
  },
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": null,
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity 200ms ease 50ms: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": {
          "type": "unit",
          "value": 50,
          "unit": "ms"
        }
      }
       1`] = `
{
  "delay": {
    "type": "unit",
    "unit": "ms",
    "value": 50,
  },
  "duration": {
    "type": "unit",
    "unit": "ms",
    "value": 200,
  },
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity 200ms ease: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
       1`] = `
{
  "delay": null,
  "duration": {
    "type": "unit",
    "unit": "ms",
    "value": 200,
  },
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity 200ms: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 200,
          "unit": "ms"
        },
        "timing": null,
        "delay": null
      }
       1`] = `
{
  "delay": null,
  "duration": {
    "type": "unit",
    "unit": "ms",
    "value": 200,
  },
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": null,
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity ease 50ms: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": {
          "type": "unit",
          "value": 50,
          "unit": "ms"
        },
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
     1`] = `
{
  "delay": null,
  "duration": {
    "type": "unit",
    "unit": "ms",
    "value": 50,
  },
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from opacity ease: 
      {
        "property": {
          "type": "keyword",
          "value": "opacity"
        },
        "duration": null,
        "timing": {
          "type": "keyword",
          "value": "ease"
        },
        "delay": null
      }
     1`] = `
{
  "delay": null,
  "duration": null,
  "property": {
    "type": "keyword",
    "value": "opacity",
  },
  "timing": {
    "type": "keyword",
    "value": "ease",
  },
}
`;

exports[`extractTransitionProperties parses and extracts values from transform 0.6s cubic-bezier(0.36, 0, 0.66, -0.56): 
      {
        "property": {
          "type": "keyword",
          "value": "transform"
        },
        "duration": {
          "type": "unit",
          "value": 0.6,
          "unit": "s"
        },
        "timing": {
          "type": "keyword",
          "value": "cubic-bezier(0.36, 0, 0.66, -0.56)"
        },
        "delay": null
      }
     1`] = `
{
  "delay": null,
  "duration": {
    "type": "unit",
    "unit": "s",
    "value": 0.6,
  },
  "property": {
    "type": "keyword",
    "value": "transform",
  },
  "timing": {
    "type": "keyword",
    "value": "cubic-bezier(0.36, 0, 0.66, -0.56)",
  },
}
`;
